zuulcilint/checker.py

```diff
@@ -121,3 +121,39 @@ def check_inexistent_nodesets(
                     inexistent_nodesets.add(job_nodeset_list["name"])
 
     return inexistent_nodesets
+
+
+def check_duplicate_semaphore(
+    jobs: list[dict | None]
+) -> set[dict[str, str] | None]:
+    """Check that when a job has a semaphore, the run entry does not have a semaphore
+    with the same name.
+
+    Args:
+    ----
+        jobs: A list of Zuul jobs.
+
+    Returns:
+    -------
+        A set of duplicated semaphores.
+    """
+    duplicate_semaphores = set()
+    for job in jobs:
+        try:
+            if isinstance(job["job"]["semaphores"], str):
+                _job_semaphore_list = [job["job"]["semaphores"]]
+            else:
+                _job_semaphore_list = job["job"]["semaphores"]
+            if isinstance(job["job"]["run"][0]["semaphores"], str):
+                _run_semaphore_list = [job["job"]["run"][0]["semaphores"]]
+            else:
+                _run_semaphore_list = job["job"]["run"][0]["semaphores"]
```

Comment:
Why assume the semaphores will always only be defined for the first entry (index 0) of the "job.run" list?

Also, this crashes if "job.run" only specifies 1 playbook path as a string, like so:
```yaml
- job:
    name: test-run-semaphores
    semaphores:
      - test-run-semaphore
    run: test.yaml
```
```
  File "...\zuulcilint\checker.py", line 147, in check_duplicate_semaphore
    if isinstance(job["job"]["run"][0]["semaphores"], str):
                  ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^
TypeError: string indices must be integers, not 'str'
```

Reply:
Fixed :)